server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java

```diff
@@ -0,0 +1,349 @@
+/*
+ *
+ *  * Copyright 2021  Red Hat, Inc. and/or its affiliates
+ *  * and other contributors as indicated by the @author tags.
+ *  *
+ *  * Licensed under the Apache License, Version 2.0 (the "License");
+ *  * you may not use this file except in compliance with the License.
+ *  * You may obtain a copy of the License at
+ *  *
+ *  * http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package org.keycloak.userprofile;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.ObjectUtil;
+import org.keycloak.models.Constants;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.utils.KeycloakModelUtils;
+import org.keycloak.userprofile.validation.AttributeValidator;
+import org.keycloak.userprofile.validation.Validator;
+
+/**
+ * <p>The default implementation for {@link Attributes}. Should be reused as much as possible by the different implementations
+ * of {@link UserProfileProvider}.
+ *
+ * <p>This implementation is not specific to any user profile implementation.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class DefaultAttributes extends HashMap<String, List<String>> implements Attributes {
+
+    private static final Logger logger = Logger.getLogger(DefaultAttributes.class);
+
+    /**
+     * There are the declarations for creating the built-in validations for read-only attributes. Regardless of the context where
+     * user profiles are used. They are related to internal attributes with hard conditions on them in terms of management.
+     */
+    private static String UPDATE_READ_ONLY_ATTRIBUTES_REJECTED = "updateReadOnlyAttributesRejectedMessage";
```

Comment:
I hope we will get rid of all these hardcoded readonly attribute definitions later as it will be based on the configuration.

Reply:
Not really. We need those as they are related to internal constraints. Regardless of the user profile configuration.

Reply:
I still believe that this hardcoded list of readonly aattributes should stay in Legacy provider only. New provider should work based on the user profile configuration - it should look at the configuration of the user profile used for actual action and treat all attributes which are NOT configured in it as readonly. And allow to modify only these which are configured for current user profile. This way you get much more predictable and configurable user profile handling.

Reply:
Sure, and that is how it works. Where only attributes defined in the profile are considered.

But again, this is something we should not remove now because it is related to existing behavior. It is expected that you can even set for instance a config to the provider for specifying addition read-only properties, with tests and etc relying on it.

Don't worry about that now as it is a implementation detail that would not impact at all user profile use cases. Which we can remove as soon as we deprecate everything related to legacy behavior.